abci/client/client.go

```diff
@@ -16,50 +15,19 @@ const (
 
 //go:generate ../../scripts/mockery_generate.sh Client
 
-// Client defines an interface for an ABCI client.
-// All `Async` methods return a `ReqRes` object.
-// All `Sync` methods return the appropriate protobuf ResponseXxx struct and an error.
-// Note these are client errors, eg. ABCI socket connectivity issues.
-// Application-related errors are reflected in response via ABCI error codes and logs.
+// Client defines the interface for an ABCI client.
+//
+// NOTE these are client errors, eg. ABCI socket connectivity issues.
+// Application-related errors are reflected in response via ABCI error codes
+// and (potentially) error response.
 type Client interface {
 	service.Service
+	types.Application
 
-	SetResponseCallback(Callback)
 	Error() error
-
-	FlushAsync() *ReqRes
-	EchoAsync(msg string) *ReqRes
-	InfoAsync(types.RequestInfo) *ReqRes
-	DeliverTxAsync(types.RequestDeliverTx) *ReqRes
-	CheckTxAsync(types.RequestCheckTx) *ReqRes
```

Comment:
Removing `CheckTxAsync` was decision that was made as part of v0.36. We have been revisiting all decisions made in v0.35 and v0.36, I think this one should be no exception.
Futher, I'm not sure I agree with `CheckTx` being synchrnous, as `CheckTx` is explained in the spec as an asynchronous method, which makes sense for performance reasons when the App is not in the same process.

Reply:
If we want `CheckTx` to be asynchronous, I'd recommend simply running it in a go routine but given that calls are serialized to the application, I'm not sure what it really provides in terms of performance.

Reply:
After doing a second pass on the beginning for this diff, I realize this goes _further_ than #7607 in removing Async methods. The original #7607 leaves the possibility of leveraging _some_ async methods in the future. This patch, which is supposed to be a backport (or a set of them), removes that possibility.